### PR TITLE
bindings: Expose the PkEncryption stuff in the crypto crate bindings

### PR DESCRIPTION
**Pull request by poljar**
Again a draft until we get things merged in vodozemac land.

@bmarty and @Velin92 I think this is all you need for the legacy apps. Please shout if there's more needed.

### CODE REVIEW

**Comment by bmarty** [comment]
@poljar I've got an issue when building this branch, not sure this is related to your change.

The command:

```sh
./scripts/build.sh -p ../matrix-rust-sdk -m crypto -t aarch64-linux-android
```

(from the project https://github.com/matrix-org/matrix-rust-components-kotlin)

and the error:

```
-- Generate uniffi files
-- library_path = /Users/bmarty/workspaces/matrix-rust-sdk/target/aarch64-linux-android/reldbg/libmatrix_sdk_crypto_ffi.so
Error: Constructor return type must be Self or Arc<Self>
```

any ideas?

**Comment by poljar** [comment]
> @poljar I've got an issue when building this branch, not sure this is related to your change.
> 
> The command:
> 
> ```shell
> ./scripts/build.sh -p ../matrix-rust-sdk -m crypto -t aarch64-linux-android
> ```
> 
> (from the project https://github.com/matrix-org/matrix-rust-components-kotlin)
> 
> and the error:
> 
> ```
> -- Generate uniffi files
> -- library_path = /Users/bmarty/workspaces/matrix-rust-sdk/target/aarch64-linux-android/reldbg/libmatrix_sdk_crypto_ffi.so
> Error: Constructor return type must be Self or Arc<Self>
> ```
> 
> any ideas?

Seems like I can't return `Option<Arc<Self>>` in the constructor, what I can return is `Result<Arc<Self>>`... Well OK, we'll add an error type.

**Comment by poljar** [comment]
@bmarty should work now, I changed the offending function to return an error instead.

**Comment by bmarty** [comment]
> @bmarty should work now, I changed the offending function to return an error instead.

Thanks, it's now compiling.
/me does the next steps.

**Comment by bmarty** [comment]
OK, the code change is in this commit, you may want to double check: https://github.com/element-hq/element-android/pull/8901/commits/355621b0c70b24e035e40ff00c7df9df334200d6

File: bindings/matrix-sdk-crypto-ffi/src/lib.rs [hunk 1]
```diff
@@ -925,6 +925,63 @@ fn vodozemac_version() -> String {
...
+}
+
+#[uniffi::export]
+impl PkEncryption {
+    /// Create a new [`PkEncryption`] object from a `Curve25519PublicKey`
+    /// encoded as Base64.
+    ///
+    /// The public key should come from an existing `PkDecryption` object.
+    /// Returns a `DecodeError` if the Curve25519 key could not be decoded
+    /// correctly..
```
Comment (jmartinesp) [inline bindings/matrix-sdk-crypto-ffi/src/lib.rs:945]: nit: `..` at the end.

**Review by jmartinesp — APPROVED** [review]
LGTM, there is a small typo in case you want to fix it.

File: bindings/matrix-sdk-crypto-ffi/src/lib.rs (hunk 1 above)
Comment (poljar) [reply bindings/matrix-sdk-crypto-ffi/src/lib.rs]: Thanks, fixed in https://github.com/matrix-org/matrix-rust-sdk/pull/3971/commits/1fa5132f4b7902cec222abab0815a80787ef6763.

**Pull request merged**